### PR TITLE
[ruff] Treat f-string interpolation as potential side effect in RUF019

### LINKED ISSUE #12953: `contains_effect` should consider f-strings to have effects
`contains_effect` should consider f-strings to have effects because of the implicit `__str__` or `__repr__` call. This affects any rule that relies on `contains_effect`; here is an example with [RUF019](https://docs.astral.sh/ruff/rules/unnecessary-key-check/).
```console
$ ruff --version
ruff 0.6.1
$ cat ruf019.py
class C:
    def __init__(self):
        self.i = 0

    def __str__(self):
        self.i += 1
        return str(self.i)


if __name__ == "__main__":
    c = C()
    dct = {"1": "a", "2": "b"}
    if f"{c}" in dct and dct[f"{c}"]:
        pass
    print(c.i)
$ python ruf019.py
2
$ ruff check --isolated --select RUF019 ruf019.py --fix
Found 1 error (1 fixed, 0 remaining).
$ python ruf019.py
1
```

Some f-strings can be statically proven to not have effects, if they contain no formatted expressions (`f"x"`) or they only format values of types formatting which is known not to have effects (`f"{1}"`).

### PR DESCRIPTION
**Pull request by anishgirianish**
## Summary

Fixes #12953                                                                                                      
                                                            
F-string interpolation can call `__format__`/`__str__`/`__repr__`, which may have side effects. RUF019 was applying a safe auto-fix that collapsed two `__str__` calls into one, changing behavior.                          
                                                                                                                    
Added a tri-state `SideEffect` enum (`No`/`Maybe`/`Yes`) and a `side_effect()` function that reuses the existing `any_over_expr` traversal via a new `FnMut` variant (`any_over_expr_mut`), following the approach suggested by @ntBre  in the other closed pr tagged in the issue

In RUF019, `SideEffect::Maybe` (non-literal f-string interpolation) now produces an unsafe fix instead of a safe one. Literal interpolations like `f"{1}"` remain safe.

## Test Plan

- Added f-string fixture cases to `RUF019.py` (non-literal → unsafe, literal → safe, no interpolation → safe).
- `cargo nextest run -p ruff_linter`                                        
- Ecosystem check (stable + preview)

### CODE REVIEW

File: crates/ruff_python_ast/src/helpers.rs [hunk 1]
```diff
@@ -57,10 +129,10 @@ where
             node_index: _,
         }) = expr
         {
-            // Ex) `list()`
             if arguments.is_empty() {
                 if let Expr::Name(ast::ExprName { id, .. }) = func.as_ref() {
                     if !is_iterable_initializer(id.as_str(), |id| is_builtin(id)) {
+                        effect = SideEffect::Yes;
                         return true;
```
Comment (MichaReiser) [inline crates/ruff_python_ast/src/helpers.rs]: Having to write both `effect = SideEffect::Yes` and `return true` feels rather fragile. 

I'm inclined to extract its own function that, given an expression, returns its side effect (could even be `SideEffect::from_expr`) and to use a `match` statement instead of the `if else` branches. Using a `match` statement reveals that we currently don't handle `Named`, which I would consider a side effect.

File: crates/ruff_linter/resources/test/fixtures/ruff/RUF019.py [hunk 2]
```diff
@@ -30,3 +30,17 @@
         d ["key"]
 ):
     ...
+
+# https://github.com/astral-sh/ruff/issues/12953
+# F-string with non-literal interpolation — unsafe fix (may invoke __str__)
+c = object()
```
Comment (MichaReiser) [inline crates/ruff_linter/resources/test/fixtures/ruff/RUF019.py]: We should use a class with a custom `__str__` method that does have side effects. This test otherwise doesn't do a good job at documenting the edge case.

File: crates/ruff_python_ast/src/helpers.rs [hunk 3]
```diff
@@ -40,6 +40,63 @@ where
...
+/// or definitely has side effects.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum SideEffect {
+    /// The expression is definitely side-effect-free.
+    No,
+    /// The expression may have side effects (e.g., f-string interpolation
+    /// may invoke `__format__` or `__str__`).
+    Maybe,
+    /// The expression definitely has side effects.
+    Yes,
```
Comment (MichaReiser) [inline crates/ruff_python_ast/src/helpers.rs]: I would rename the variants to `Absent`, `Possible`, `Present` or `None`, `Uncertain`, `Definite`. They seem clearer then `Yes` and `No`.

File: crates/ruff_python_ast/src/helpers.rs [hunk 4]
```diff
@@ -106,13 +179,28 @@ where
...
+            }
+            return false;
+        }
+
         // Otherwise, avoid all complex expressions.
-        matches!(
+        if matches!(
             expr,
             Expr::Await(_)
                 | Expr::Call(_)
```
Comment (MichaReiser) [inline crates/ruff_python_ast/src/helpers.rs]: It feels a bit confusing now that, e.g., `Call` returns `Yes` even though we don't know for sure. We simply assume it has side effects because we don't know better.

File: crates/ruff_python_ast/src/helpers.rs (hunk 1 above)
Comment (anishgirianish) [reply crates/ruff_python_ast/src/helpers.rs]: Refactored to SideEffect::from_expr with a match. Added Named as well. thankss

File: crates/ruff_linter/resources/test/fixtures/ruff/RUF019.py (hunk 2 above)
Comment (anishgirianish) [reply crates/ruff_linter/resources/test/fixtures/ruff/RUF019.py]: Updated the fixture to use a class with a side-effectful __str__. Thanks

File: crates/ruff_python_ast/src/helpers.rs (hunk 3 above)
Comment (anishgirianish) [reply crates/ruff_python_ast/src/helpers.rs]: Renamed to Absent/Possible/Present. Thanks

File: crates/ruff_python_ast/src/helpers.rs (hunk 4 above)
Comment (anishgirianish) [reply crates/ruff_python_ast/src/helpers.rs]: I think the rename to Present helps here since it reads as 'we treat it as present' rather than implying certainty. Thanks

**Comment by anishgirianish** [comment]
@MichaReiser  Thank you so much for the review! Addressed all feedback, refactored to SideEffect::from_expr with a match, renamed variants to Absent/Possible/Present, added Named as a side effect, and updated the fixture with a class that has a side-effectful __str__. I really appreciate the suggestions. Ready for re-review whenever you have a moment.

File: crates/ruff_python_ast/src/helpers.rs [hunk 5]
```diff
@@ -40,6 +40,145 @@ where
...
+            (Self::Possible, _) | (_, Self::Possible) => Self::Possible,
+            _ => Self::Absent,
+        }
+    }
+
+    /// Classify a single expression node's side effect.
+    ///
+    /// Returns `Some(effect)` if this node determines the side effect level,
+    /// or `None` to continue walking child expressions.
+    fn from_expr(expr: &Expr, is_builtin: &dyn Fn(&str) -> bool) -> Option<Self> {
```
Comment (MichaReiser) [inline crates/ruff_python_ast/src/helpers.rs]: Do we need the `Option` here, given that `Self::Possible` and `::Absent` both continue walking the subtree? Can we return `Absent` in the cases where we're returning `None` right now?

File: crates/ruff_python_ast/src/helpers.rs [hunk 6]
```diff
@@ -40,6 +40,145 @@ where
...
+            | Expr::DictComp(_)
+            | Expr::Generator(_)
+            | Expr::ListComp(_)
+            | Expr::SetComp(_)
+            | Expr::Subscript(_)
+            | Expr::Yield(_)
+            | Expr::YieldFrom(_)
+            | Expr::IpyEscapeCommand(_) => Some(Self::Present),
+
+            _ => None,
```
Comment (MichaReiser) [inline crates/ruff_python_ast/src/helpers.rs]: I suggest making this an exhaustive match, to ensure we think about `contains_effect` when adding a new expression type.

File: crates/ruff_python_ast/src/helpers.rs [hunk 7]
```diff
@@ -40,6 +40,145 @@ where
...
+        }
+    }
+}
+
+fn is_known_safe_binop_operand(expr: &Expr) -> bool {
```
Comment (MichaReiser) [inline crates/ruff_python_ast/src/helpers.rs]: Nit: You could do the same for the other match methods.
```suggestion
const fn is_known_safe_binop_operand(expr: &Expr) -> bool {
```

File: crates/ruff_python_ast/src/helpers.rs (hunk 5 above)
Comment (anishgirianish) [reply crates/ruff_python_ast/src/helpers.rs]: Done, returning directly now. Thank you

File: crates/ruff_python_ast/src/helpers.rs (hunk 6 above)
Comment (anishgirianish) [reply crates/ruff_python_ast/src/helpers.rs]: Done, updated to use exhaustive match. thank you

File: crates/ruff_python_ast/src/helpers.rs (hunk 7 above)
Comment (anishgirianish) [reply crates/ruff_python_ast/src/helpers.rs]: Done, updated both to use exhaustive match and const fn. thank you

**Comment by anishgirianish** [comment]
@MichaReiser thank you so much for the review. Addressed all three feedback, removed Option from from_expr, made both matches exhaustive, added const fn.  Ready for another look whenever you get a chance.

thank you
